### PR TITLE
Enhance Tap-sftp to Support Multiple Encoding Formats

### PR DESCRIPTION
**Pull request by sgandhi1311**
# Description of change
Currently, Tap-sftp exclusively supports the utf-8 encoding format. In order to enhance its versatility and compatibility, we are planning to expand support to encompass all encoding formats specified by Python 3.9.

# Manual QA steps
 - Done the manual testing by passing different encoding formats.
 - Kept the encoding format blank in the configuration, to test it sets the default to _utf-8_
 
# Risks
 - Low
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Please add unit tests.

**Comment by sgandhi1311** [comment]
> Please add unit tests.

done

File: tap_sftp/helper.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+def write_message(message, ensure_ascii=True):
+    sys.stdout.write(format_message(message, ensure_ascii=ensure_ascii) + '\n')
+    sys.stdout.flush()
+
+
+def write_record(stream_name, record, stream_alias=None, time_extracted=None, ensure_ascii=True):
+    """Write a single record for the given stream.
+    
+    """
+    write_message(RecordMessage(stream=(stream_alias or stream_name),
```
Comment (RushiT0122) [inline tap_sftp/helper.py:19]:
```suggestion
def write_record(stream_name, record, stream_alias=None, time_extracted=None, ensure_ascii=True):
    """
    Write a single record for the given stream.
    
    """
    write_message(RecordMessage(stream=(stream_alias or stream_name),
```

File: tap_sftp/sync.py [hunk 2]
```diff
@@ -81,7 +87,8 @@ def sync_file(conn, f, stream, table_spec):
 
                 to_write = transformer.transform(rec, stream.schema.to_dict(), metadata.to_map(stream.metadata))
 
-                singer.write_record(stream.tap_stream_id, to_write)
+                write_record(stream.tap_stream_id, to_write, ensure_ascii=False)
+                # singer.write_record(stream.tap_stream_id, to_write)
```
Comment (RushiT0122) [inline tap_sftp/sync.py:91]: Remove this comment.

File: tests/unittests/test_encoding_format.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        try:
+            with patch("sys.stdout", new_callable=StringIO) as mock_stdout:
+                do_discover(config)
+                output = mock_stdout.getvalue().strip()
+                expected_output = json.dumps(
+                    {"streams": ["stream1", "stream2"]}, indent=2
+                )
+                self.assertEqual(output, expected_output)
+        except Exception as e:
+            self.fail(f"Exception occurred: {str(e)}")
```
Comment (RushiT0122) [inline tests/unittests/test_encoding_format.py:49]: Some queries:
- Are we expecting any exception to occur in this test?
- At what point we are expecting exception to occur? 
- Shouldn't put try-catch block around it and use appropriate assertions to validate the exception?

File: tests/unittests/test_permission_error.py [hunk 4]
```diff
@@ -68,7 +69,7 @@ def test_permisison_error_during_sync(self, mocked_get_row_iterators, mocked_log
 
         conn = client.SFTPConnection("10.0.0.1", "username", port="22")
 
-        rows_synced = sync.sync_file(conn, {"filepath": "/root_dir/file.csv.gz", "last_modified": "2020-01-01"}, None, {"key_properties": ["id"], "delimiter": ","})
+        rows_synced = sync.sync_file(conn, {"filepath": "/root_dir/file.csv.gz", "last_modified": "2020-01-01"}, None, {"key_properties": ["id"], "delimiter": ","}, encoding_format=DEFAULT_ENCODING_FORMAT)
```
Comment (RushiT0122) [inline tests/unittests/test_permission_error.py:72]: Line is too long with addition of encoding, please refactor it here and at other location as well.

File: tests/test_sftp_encoding.py [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+        # assert that each of the streams that we synced are the ones that we expect to see
+        record_count_by_stream = runner.examine_target_output_file(self,
+                                                                   conn_id,
+                                                                   self.expected_first_sync_streams(),
+                                                                   self.expected_pks())
+
+        # Verify that the full table was syncd
+        for tap_stream_id in self.expected_first_sync_streams():
+            self.assertEqual(self.expected_first_sync_row_counts()[tap_stream_id],
+                             record_count_by_stream[tap_stream_id])
```
Comment (RushiT0122) [inline tests/test_sftp_encoding.py:213]: We should validate the extracted records with actual source records.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Requested changes in-line.

File: tests/unittests/test_encoding_format.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+            )
+            self.assertEqual(output, expected_output)
+
+        if encoding_format == "utf-8":
+            # Bypassing encoding check for `utf-8` as it is widely used
+            mock_is_valid_encoding.assert_not_called()
+        else:
+            mock_is_valid_encoding.assert_called_with("latin_1")
+        mock_discover_streams.assert_called_with(config, encoding_format)
+        self.assertEqual(captured_output, sys.stdout)  # Ensure sys.stdout is restored
```
Comment (RushiT0122) [inline tests/unittests/test_encoding_format.py:54]: We should not use if-else for assertion specially in parameterised tests. We should separate these tests.

**Review by RushiT0122 — APPROVED** [review]

File: tap_sftp/__init__.py [hunk 7]
```diff
@@ -4,17 +4,24 @@
...
+DEFAULT_ENCODING_FORMAT = "utf-8"
 
 def do_discover(config):
     LOGGER.info("Starting discover")
-    streams = discover_streams(config)
+    # validate the encoding format
+    encoding_format = config.get("encoding_format") or DEFAULT_ENCODING_FORMAT
+    if encoding_format != "utf-8":
+        if not is_valid_encoding(encoding_format):
+            raise Exception("Unknown Encoding - {}. Enter the valid encoding format".format(encoding_format))
```
Comment (cosimon) [inline tap_sftp/__init__.py:23]: `is_valid_encoding()` returns True for `utf-8`, these if statements can be combined.

Comment (sgandhi1311) [reply tap_sftp/__init__.py]: combined the if statements.

**Pull request merged**